src/com/reason/ide/console/CompilerAction.java

```diff
@@ -27,31 +24,63 @@
-                    VirtualFile baseDirectory = baseDirectoryOptional.get();
-                    console.print("No active text editor found, using root directory " +
-                                    baseDirectory.getPath() + "\n", ConsoleViewContentType.NORMAL_OUTPUT);
-                    VirtualFile bsConfig = baseDirectory.findChild(BsConstants.BS_CONFIG_FILENAME);
-                    assert bsConfig != null;
-                    compiler.run(bsConfig, cliType, null);
-                }
-            }
+        Optional<Editor> editorOptional = getActiveEditor(project);
+        if (editorOptional.isPresent()) {
+            compileFile(project, editorOptional.get(), cliType);
+            return;
+        }
+        compileDirectory(project, cliType);
+    }
+
+    private static void compileDirectory(@NotNull Project project, CliType cliType) {
+        ORCompilerManager compilerManager = ORCompilerManager.getInstance(project);
+        Optional<Compiler> compilerOptional = compilerManager.getCompiler(cliType);
+        if (!compilerOptional.isPresent()) {
+           return;
+        }
+        Compiler compiler = compilerOptional.get();
+        ConsoleView consoleView = compiler.getConsoleView();
+        if (consoleView == null) {
+            return;
+        }
+        Optional<VirtualFile> baseDir = compiler.findFirstContentRoot(project);
+        if (!baseDir.isPresent()) {
+            consoleView.print("Can't find content root\n", ConsoleViewContentType.NORMAL_OUTPUT);
         } else {
-            Document document = editor.getDocument();
-            PsiFile psiFile = PsiDocumentManager.getInstance(project).getPsiFile(document);
-            if (psiFile != null) {
-                compiler.run(psiFile.getVirtualFile(), cliType, null);
-            }
+            consoleView.print("No active text editor found, using root directory " + baseDir.get().getPath() + "\n",
+                    ConsoleViewContentType.NORMAL_OUTPUT);
+            compiler.run(baseDir.get(), cliType, null);
+        }
+    }
+
+    private static void compileFile(@NotNull Project project, @NotNull Editor editor, @NotNull CliType cliType) {
+        Optional<PsiFile> activeFileOptional = getActiveFile(project, editor);
+        if (!activeFileOptional.isPresent()) {
+            return;
+        }
+        PsiFile activeFile = activeFileOptional.get();
+        // unsupported file type is open, compile the directory instead
+        if (!Compiler.isSupportedFileType(activeFile.getFileType())) {
+            compileDirectory(project, cliType);
+            return;
         }
+        ORCompilerManager compilerManager = ORCompilerManager.getInstance(project);
+        Optional<Compiler> compilerOptional = compilerManager.getCompiler(cliType);
+        if (!compilerOptional.isPresent()) {
+            return;
+        }
+        Compiler compiler = compilerOptional.get();
+        compiler.run(activeFile.getVirtualFile(), cliType, null);
```

Comment:
if optionals are used, we could streamline the commands, like in
```
        ORCompilerManager.getInstance(project).
                getCompiler(cliType).
                ifPresent(compiler -> compiler.run(activeFile.getVirtualFile(), cliType, null));
```

Reply:
at that point, we have already checked for the presence of the optional above so another check is redundant. i do prefer the functional style though.